diesel/src/row.rs

```diff
@@ -1,30 +1,164 @@
 //! Contains the `Row` trait
 
-use crate::backend::{self, Backend};
-use crate::deserialize::{self, FromSql};
+use crate::{
+    backend::{self, Backend},
+    deserialize,
+};
+use deserialize::FromSql;
+use std::ops::Range;
+
+/// Representing a way to index into database rows
+///
+/// * Crates using existing backends should use existing implementations of
+///   this traits. Diesel provides `RowIndex<usize>` and `RowIndex<&str>` for
+///   all bulit-in backends
+///
+/// * Crates implementing custom backends need to provide `RowIndex<usize>` and
+///   `RowIndex<&str>` impls for their [`Row`] type.
+///
+/// [`Row`]: trait.Row.html
+pub trait RowIndex<I> {
+    /// Get the numeric index inside the current row for the provided index value
+    fn idx(&self, idx: I) -> Option<usize>;
+}
 
 /// Represents a single database row.
-/// Apps should not need to concern themselves with this trait.
 ///
-/// This trait is only used as an argument to [`FromSqlRow`].
+/// This trait is used as an argument to [`FromSqlRow`].
 ///
 /// [`FromSqlRow`]: ../deserialize/trait.FromSqlRow.html
-pub trait Row<DB: Backend> {
-    /// Returns the value of the next column in the row.
-    fn take(&mut self) -> Option<backend::RawValue<DB>>;
+pub trait Row<'a, DB: Backend>: RowIndex<usize> + for<'b> RowIndex<&'b str> + Sized {
+    /// Field type returned by a `Row` implementation
+    ///
+    /// * Crates using existing backend should not concern themself with the
+    ///   concrete type of this associated type.
+    ///
+    /// * Crates implementing custom backends should provide their own type
+    ///   meeting the required trait bounds
+    type Field: Field<'a, DB>;
 
-    /// Returns whether the next `count` columns are all `NULL`.
+    /// Return type of `PartialRow`
     ///
-    /// If this method returns `true`, then the next `count` calls to `take`
-    /// would all return `None`.
-    fn next_is_null(&self, count: usize) -> bool;
-
-    /// Skips the next `count` columns. This method must be called if you are
-    /// choosing not to call `take` as a result of `next_is_null` returning
-    /// `true`.
-    fn advance(&mut self, count: usize) {
-        for _ in 0..count {
-            self.take();
+    /// For all implementations, beside of the `Row` implementation on `PartialRow` itself
+    /// this should be `Self`.
+    #[doc(hidden)]
+    type InnerPartialRow: Row<'a, DB>;
+
+    /// Get the number of fields in the current row
+    fn field_count(&self) -> usize;
+
+    /// Get the field with the provided index from the row.
+    ///
+    /// Returns `None` if there is no matching field for the given index
+    fn get<I>(&self, idx: I) -> Option<Self::Field>
+    where
+        Self: RowIndex<I>;
+
+    /// Returns a wrapping row that allows only to access fields, where the index is part of
+    /// the provided range.
+    #[doc(hidden)]
```

Comment:
I do not see how this would be relevant for dynamic queries. They normally should use `RowIndex<&str>` or just load all columns that are queried.

Reply:
I could imagine a scenario where there is still some form of known hierarchical structure, but user-defined and unknown at compile time which could use this when building it through a recursive function.
What would be the downside of having it public? (with corresponding documentation explaining this could only be useful when implementing custom parsing)

Reply:
I think we can make this function public at any later point if someone comes up with a concrete use case.